src/librustc/traits/mod.rs

```diff
@@ -242,6 +244,59 @@ pub type Obligations<'tcx, O> = Vec<Obligation<'tcx, O>>;
 pub type PredicateObligations<'tcx> = Vec<PredicateObligation<'tcx>>;
 pub type TraitObligations<'tcx> = Vec<TraitObligation<'tcx>>;
 
+#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
+pub enum WhereClauseAtom<'tcx> {
+    Implemented(ty::PolyTraitPredicate<'tcx>),
+    ProjectionEq(ty::PolyProjectionPredicate<'tcx>),
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
+pub enum DomainGoal<'tcx> {
+    Holds(WhereClauseAtom<'tcx>),
+    WellFormed(WhereClauseAtom<'tcx>),
+    FromEnv(WhereClauseAtom<'tcx>),
+    WellFormedTy(Ty<'tcx>),
+    FromEnvTy(Ty<'tcx>),
+    RegionOutlives(ty::PolyRegionOutlivesPredicate<'tcx>),
+    TypeOutlives(ty::PolyTypeOutlivesPredicate<'tcx>),
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
+pub enum LeafGoal<'tcx> {
+    DomainGoal(DomainGoal<'tcx>),
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
+pub enum QuantifierKind {
+    Universal,
+    Existential,
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub enum Goal<'tcx> {
+    Implies(Vec<DomainGoal<'tcx>>, Box<Goal<'tcx>>),
```

Comment:
So, this matches chalk, where we make the hypotheses a set of domain goals, but it's not as general as we could do. We really should use `Clause<'tcx>` here, which would match the following enum:

```rust
enum Clause<'tcx> {
    Implies(Vec<Goal<'tcx>>, DomainGoal<'tcx>),
    DomainGoal(DomainGoal<'tcx>),
    ForAll(Box<ty::Binder<Clause<'tcx>>),
}
```

This matches the definition from Page 7 of "A Proof Procedure for the Logic of Hereditary Harrop Formulas" pretty well. That definition is `D = A | G => A | D ^ D | forall x. D` -- where `D` is a clause, `G` is a goal, and `A` is a `DomainGoal` -- but we've moved the "and" case (`D ^ D`) into the vector. (There are broader definitions as well for clauses, but they can be normalized into this form.)

This would also entail a change to `Environment` to store these clauses instead of just domain goals.

The reason that this matters is that if we move the binder out of `DomainGoal`, we need somewhere to put it in the case of a clause!

Reply:
Also, I think we should use interners probably, and hence make this:

```rust
type Goal<'tcx> = &'tcx GoalKind<'tcx>;
enum GoalKind<'tcx> {
    Implies(Vec<DomainGoal<'tcx>, Goal<'tcx>),
    ... // as today
}
```

but we can do that in a follow-up. (We should open a FIXME for it before landing.)

Reply:
I opened https://github.com/rust-lang-nursery/chalk/issues/94 to describe the change here in chalk